### ISSUE TITLE
TODO - Expected developments

### ISSUE THREAD

**Issue by Emory-NLP** [NLP-Suite/NLP-Suite#44]
1.	**Excel**
a.	Rewrite some of the functions that prepare csv data for visualization in Excel.
b.	Frequency & normalized values (percentages)
Frequency calculation (e.g., Excel charts) should export both absolute frequencies and normalized frequencies (namely, percentages); normalized frequencies are a must when considering a multiple file corpus so that the results become comparable.
c.	Allow the selection of specific file(s) for line charts when processing merged csv files containing multiple documents.
5.	**Search and split options**
Several GUIs (file_finder_byWord_main.py, file_splitter_main.py,  sentence_analysis_main) have various ways of searching for words/collocations and splitting text. Uniform all search and split options by word/collocations with similar code in file_splitter_ByKeyword_txt_util.py, sentence_analysis_util.extract_sentences); make sure they all have the same code behind and the same GUI widget options.
6.	**Hedge annotator**
Add available Python hedge/uncertainty annotators to the NLP Suite
https://github.com/PAJEAN/uncertaintyDetection#readme
https://github.com/meyersbs/uncertainty 
7.	**Spelling checker**
a.	Rewrite the current spelling checker available in the NLP Suite that relies on NLTK and use instead fuzzywuzy
b.	Check the performance of pyspellchecker, autocorrect, textblob
8.	**Grammar checker**
https://github.com/viraja1/grammar-check 
https://stackoverflow.com/questions/10252448/how-to-check-whether-a-sentence-is-correct-simple-grammar-check-in-python
https://sourcecontribute.com/2017/11/06/languagetool-free-grammar-checking-tool/ 
9.	**style_analysis_main.py**
a.	Must complete the script with csv output files.
b.	**For Yule K, export both I and K and the number of distinct values.
10.	**narrative_analysis_main.py**
Complete the pipeline
11.	**whats_in_your_corpus**
connect the nature reference
12.	**classifiers, word embedding, Subject-Verb-Object (SVO) extraction, and Semantic Role Labeling (SRL)** 
Complete the work using:
a.	Connect the TensorFlow script in the Suite to a GUI
b.	Classifiers and K-Nearest Neighbor (https://scikit-learn.org/stable/modules/neighbors.html)
d.	SVO/SRL looking at Senna (http://ronan.collobert.com/senna/), FreeLing (https://freeling-tutorial.readthedocs.io/en/latest/code/example05.py/), FrameNet (https://framenet.icsi.berkeley.edu/fndrupal/ASRL), Illinois
(https://cogcomp.cs.illinois.edu/page/software_view/SRL; demo http://cogcomp.cs.illinois.edu/page/demo_view/srl), Edinburgh (http://homepages.inf.ed.ac.uk/mroth/demo.html). 
13.	**Mallet**
Take the list of words per topic and extract all the sentences that contain those words.
15.	**GIS**
a.	Complete the pipeline in GIS_main.py go through the GIS_pipeline
b.	Connect GIS to QGIS
c.	Connect the Python heatmap designed by Denny
d.	Complete the GIS_distance_util.
16.	**Pandas**
a.	We have written the function df_to_csv (similar to the list_to_csv function) that checks if a csv file is open with Permission denied error in command line; the df_to_csv would dislay the error as a more user-friendly tkinter message. The df_to_csv function must be invoked everywhere we save with pandas (for an example, see saveReminder…). 
b.	Deal with the Pandas warning
17.	**Lemmatizer**
Many of the NLP Suite scripts rely on NLTK for a lemmatizer. The NLTK lemmatizer does not perform very well. We need to explore different options. See here a list of lemmatizers:
https://www.machinelearningplus.com/nlp/lemmatization-examples-python/ 
The author provides the code for different lemmatization methods, including CoreNLP 
For spacy see also https://medium.com/@makcedward/nlp-pipeline-lemmatization-part-3-4bfd7304957 
18.	**CoReference resolution**
The NLP Suite relies on Stanford CoreNLP for Coreference resolution. Perhaps we can explore using Spacy neural network approach (https://github.com/huggingface/neuralcoref). See the comments: NeuralCoref is accompanied by a visualization client NeuralCoref-Viz, a web interface powered by a REST server that can be tried online (https://spacy.io/universe/project/neuralcoref).
20.	**Document similarities**
The NLP Suite relies on the Java script Lucene.jar to check for document similarities. We should explore other options, such as SequenceMatcher in the difflib python library (https://docs.python.org/2/library/difflib.html).
If seq.ratio() is slow, use seq.quick_ratio()
21.	**DB SQLite** 
a.	Write TIPS on how it works
b.	Further explore the option of how to visualize a database structure
c.	Write the standard SQL queries tat would allow a user immediate searches
22.	**Hover-over effects in tkinter**
Perhaps introduce hover over on some of the tkinter GUI widgets
https://stackoverflow.com/questions/20399243/display-message-when-hovering-over-something-with-mouse-cursor-in-python 
23.	**pdf conversion**
Check out the best Python libraries https://towardsdatascience.com/pdf-preprocessing-with-python-19829752af9f and perhaps extend our reliance on pdfminer
